ctx.py

```diff
@@ -12,21 +12,102 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import enum
 import functools
+import os
 
+from pathlib import Path
+import util
 '''
 Execution context. Filled upon invocation of cli.py, read by submodules
 '''
 
 args=None # the parsed command line arguments
 
-def _cfg_factory_from_dir():
-    # XXX: args does always have a cfg_dir attribute, but pylint does not always understand this
-    if not args or not hasattr(args, 'cfg_dir') or not getattr(args, 'cfg_dir'):
-        return None
+class ConfigKey(enum.Enum):
+    CONTEXT = 'ctx'
+    TERMINAL = 'terminal'
+
+    @classmethod
+    def has_key(cls, key: str):
+        return key in [e.value for e in cls]
+
+
+class ContextConfig(object):
```

Comment:
please restructure:

- one type (/class) per cfg element (ContextConfig + TerminalConfig)
- inherit from concourse.pipelines.modelbase.ModelBase
- implement getters (w/o get-prefix, please ;-)) for known attributes (maybe no setters required, see next)
- make clever use of _defaults_dict() from ModelBase to inject base / initial values


maybe even better:

implement factory methods (1 for ctx, 1 for terminal):
- collect defaults dict
- collect dict from env
- collect dict from args
- ...
-> collect those in a sequence
--> calculate effective cfg using util.merge_dicts